arangod/Scheduler/Scheduler.cpp

```diff
@@ -32,674 +32,149 @@
 
-  while (true) {
-    double const now = TRI_microtime();
+    while (_cronQueue.size() > 0) {
+      // top is a reference to a tuple containing the timepoint and a shared_ptr to the work item
+      auto const& top = _cronQueue.top();
 
-    {
-      MUTEX_LOCKER(locker, _threadCreateLock);
-
-      uint64_t const counters = _counters.load();
-
-      if (isStopping(counters)) {
-        // do not start any new threads in case we are already shutting down
-        break;
-      }
-
-      uint64_t const nrRunning = numRunning(counters);
-      uint64_t const nrWorking = numWorking(counters);
-
-      // some threads are idle
-      if (nrWorking < nrRunning) {
-        break;
-      }
-
-      // all threads are maxed out
-      _lastAllBusyStamp = now;
+      if (top.first < now) {
+        // It is time to scheduler this task, try to get the lock and obtain a shared_ptr
+        // If this fails a default WorkItem is constructed which has disabled == true
+        auto item = top.second.lock();
+        if (item) {
+          item->run();
+        }
+        _cronQueue.pop();
+      } else {
+        auto then = (top.first - now);
 
-      // reached maximum
-      if (nrRunning >= _maxThreads) {
-        break;
+        sleepTime = (sleepTime > then ? then : sleepTime);
+        break ;
       }
-
-      // increase nrRunning by one here already, while holding the lock
-      incRunning();
     }
 
-    // create thread and sleep without holding the mutex
-    try {
-      startNewThread();
-    } catch (...) {
-      // cannot create new thread or start new thread
-      // if this happens, we have to rollback the increase of nrRunning again
-      {
-        MUTEX_LOCKER(locker, _threadCreateLock);
-        decRunning();
-      }
-      // add an extra sleep so the system has a chance to recover and provide
-      // the needed resources
-      std::this_thread::sleep_for(std::chrono::microseconds(20000));
-    }
+    _croncv.wait_for(guard, sleepTime);
 
-    std::this_thread::sleep_for(std::chrono::microseconds(5000));
   }
 }
 
-void Scheduler::threadHasStopped() {
-  MUTEX_LOCKER(locker, _threadCreateLock);
-  decRunning();
-}
-
-bool Scheduler::threadShouldStop(double now) {
-  // make sure no extra threads are created while we check the timestamp
-  // and while we modify nrRunning
-
-  MUTEX_LOCKER(locker, _threadCreateLock);
+Scheduler::WorkHandle Scheduler::queueDelay(
+    RequestLane lane,
+    clock::duration delay,
+    std::function<void(bool cancelled)> handler
+) {
 
-  // fetch all counters in one atomic operation
-  uint64_t counters = _counters.load();
-  uint64_t const nrRunning = numRunning(counters);
+  TRI_ASSERT(!isStopping());
 
-  if (nrRunning <= _minThreads) {
-    // don't stop a thread if we already reached the minimum
-    // number of threads
-    return false;
+  if (delay < std::chrono::milliseconds(1)) {
+    // execute directly
+    queue(lane, [handler](){ handler(false); });
```

Comment:
A copy of the `handler` is queued directly here and then another in the `_cronQueue` a few lines down. Is this intentional? This causes the handler to be run twice.